### PR TITLE
fix: number serialization

### PR DESCRIPTION
**Pull request by markflorkowski**
Address issues with passing unserializable numbers to prepareUpload (eg. maxFileCount: Infinity)

https://discord.com/channels/966627436387266600/1102510616326967306/1276614012825767966

### CODE REVIEW

File: packages/shared/src/utils.ts [hunk 1]
```diff
@@ -79,6 +79,14 @@ export const fillInputRouteConfig = (
...
 
     newConfig[key] = { ...defaultValues, ...value };
+
+    if (newConfig[key].maxFileCount === Infinity) {
+      newConfig[key].maxFileCount = Number.MAX_SAFE_INTEGER;
+    }
+
+    if (newConfig[key].minFileCount === Infinity) {
+      newConfig[key].minFileCount = Number.MAX_SAFE_INTEGER;
+    }
```
Comment (juliusmarminge) [inline packages/shared/src/utils.ts:89]: these are unnecessary now right?

File: packages/shared/src/utils.ts [hunk 2]
```diff
@@ -276,3 +284,26 @@ export const resolveMaybeUrlArg = (maybeUrl: string | URL | undefined): URL => {
...
+/**
+ * Replacer for JSON.stringify that will replace numbers that cannot be
+ * serialized to JSON with "reasonable equivalents".
+ *
+ * Infinity and -Infinity are replaced by MAX_SAFE_INTEGER and MIN_SAFE_INTEGER
+ * NaN is replaced by 0
+ *
+ */
+export const safeNumberReplacer = (_: string, value: unknown) => {
+  if (typeof value === "number") {
```
Comment (juliusmarminge) [inline packages/shared/src/utils.ts:297]: nit: prefer early return

File: packages/shared/src/utils.ts [hunk 3]
```diff
@@ -276,3 +276,26 @@ export const resolveMaybeUrlArg = (maybeUrl: string | URL | undefined): URL => {
...
+ * Replacer for JSON.stringify that will replace numbers that cannot be
+ * serialized to JSON with "reasonable equivalents".
+ *
+ * Infinity and -Infinity are replaced by MAX_SAFE_INTEGER and MIN_SAFE_INTEGER
+ * NaN is replaced by 0
+ *
+ */
+export const safeNumberReplacer = (_: string, value: unknown) => {
+  if (typeof value !== "number") return value;
+  else {
```
Comment (juliusmarminge) [inline packages/shared/src/utils.ts:290]: But: Redundant  else

**Review by juliusmarminge — APPROVED** [review]

File: packages/uploadthing/src/internal/handler.ts [hunk 4]
```diff
@@ -357,13 +358,16 @@ const handleUploadAction = Effect.gen(function* () {
...
     {
       method: "POST",
-      body: JSON.stringify({
-        files: filesWithCustomIds,
-        routeConfig: parsedConfig,
-        metadata,
-        callbackUrl: callbackUrl.origin + callbackUrl.pathname,
-        callbackSlug: opts.slug,
-      }),
+      body: JSON.stringify(
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:361]: is this the only place we need to use this replacer? how bout when sending route config to client?

Comment (markflorkowski) [reply packages/uploadthing/src/internal/handler.ts:361]: Ah good catch

File: packages/react/src/next-ssr-plugin.tsx [hunk 5]
```diff
@@ -18,7 +19,7 @@ export function NextSSRPlugin(props: { routerConfig: EndpointMetadata }) {
   useServerInsertedHTML(() => {
     const html = [
       // Hydrate routerConfig on client globalThis
-      `globalThis.__UPLOADTHING = ${JSON.stringify(props.routerConfig)};`,
+      `globalThis.__UPLOADTHING = ${JSON.stringify(props.routerConfig, safeNumberReplacer)};`,
```
Comment (markflorkowski) [inline packages/react/src/next-ssr-plugin.tsx:22]: may cause mismatch?

Comment (juliusmarminge) [reply packages/react/src/next-ssr-plugin.tsx:22]: Why if the fetch call also uses the replacer?

File: packages/uploadthing/src/server.ts
Comment (juliusmarminge) [inline packages/uploadthing/src/server.ts:1]: What about the other adapters?

Maybe it's better to handle this replacer earlier when parsing the route config?

Comment (markflorkowski) [reply packages/uploadthing/src/server.ts:1]: Like I was originally doing in `fillInputRouteConfig`?

Comment (juliusmarminge) [reply packages/uploadthing/src/server.ts:1]: Oh maybe 😅

Maybe like `JSON.parse(JSON.stringify(data, replacer))` when sanitizing the input 🤔

Comment (markflorkowski) [reply packages/uploadthing/src/server.ts:1]: image
?

Comment (juliusmarminge) [reply packages/uploadthing/src/server.ts:1]: 👍 that should avoid having to touch all adapters right?

Comment (markflorkowski) [reply packages/uploadthing/src/server.ts:1]: I believe so, yeah.

File: packages/react/src/next-ssr-plugin.tsx (hunk 5 above)
Comment (juliusmarminge) [inline packages/react/src/next-ssr-plugin.tsx:22]: i think this is redundant given both the GET endpoint and the `extractRouterCongi` (which is what you're meant to pass as props here) use `buildPermissionHandler` which in turn invokes `fillRouterInput...` which does the replacer - so it should be all valid JSON

**Review by juliusmarminge — APPROVED** [review]

File: packages/react/src/next-ssr-plugin.tsx (hunk 5 above)
Comment (markflorkowski) [reply packages/react/src/next-ssr-plugin.tsx:22]: You're right, I just forgot about it 😅

**Pull request merged**